### ISSUE TITLE
No server matching '${pathname}' path defined in OpenAPI spec 

### ISSUE THREAD

**Issue by GSDTN** [RuntimeTools/chai-openapi-response-validator#30]
I am currently using an axios response for the request:
```js
const res = await axios.get(host+testUrl, {"headers": headers} );
```
When I run `expect(res).to.satisfyApiSpec`; I get the error `No server matching path as defined in Open API Spec.`

The reason appears to be that the axios `req.path` does not include the host portion, only the URL portion, and since it is comparing against host portions in the server list it fails and the whole test fails to move on.

I was able to 'bypass' this issue by setting a server in my Open API spec to / (forward slash).  Since that is the first part of the URL portion.  However, I would prefer not to have that in my Spec.

I was also able to get things to work if I commented out the Server check.  But as I try and push this into pipeline builds etc it would require that I also deploy all the NodeJS modules.  If there is ever a change to this package I would have to re-comment out the code.

I am not convinced the Server check is necessary (as my tests worked great when it was commented out) - so can that be completely bypassed, or at least have some configuration option to bypass the check.
Alternatively, check the request.path to see if it even resembles a domain before checking.  Look for http or periods.
Thanks

**Comment by rwalle61** [comment]
Thanks for raising this, I agree with the need to support servers in the form of https://myapi.mycompany.com. 

I actually started a branch to do this after resolving https://github.com/RuntimeTools/chai-openapi-response-validator/issues/17, but got sidetracked as no one was asking for it. Now that you (and presumably others) want it, I can look at this again. 

I'll take https://github.com/RuntimeTools/chai-openapi-response-validator/issues/17#issuecomment-565535245 into account. 

I've only briefly looked at this, but it's interesting that you were able to 
> 'bypass' this issue by setting a server in my Open API spec to / (forward slash)

Our [code does seem to expect a `server` object](https://github.com/RuntimeTools/chai-openapi-response-validator/blob/master/lib/classes/OpenApi3Spec.js#L8). Perhaps the fix involves changing that. 

Or, the [OpenAPI 3 docs say that](https://swagger.io/docs/specification/api-host-and-base-path/) 
>If the servers array is not provided or is empty, the server URL defaults to /:

    servers:
      - url: /

so perhaps we should just ensure that the openapi 3 object has at least a `servers: [{ url: '/' }]` object.

I may have time next weekend or after Christmas, or can review PRs before then :)

**Comment by GSDTN** [comment]
In many cases, I think the server list will contain one or more servers.  In our case, we provide both a production and a test server.  From the generated docs, than Try Out and Execute go against the selected server.
The other challenge here, is that the request.path does not include the server portion, only the URL portion.
From the previous ticket, it appears that the problem was that a 'base' URL was being set in the server list, and than not being considered when looking for the path.  Since it is possible that others may also want to put {server}/{basePath} in the server list, it might just be that you need to inspect the server list and strip out the protocol://host.domain:port/ sections and than use the rest of the path as a prefix for the paths being compared to the request.path.

I tried pulling this project to create a PR - but it appears it is created on a linux platform and I am on a Windows platform and there were a bunch of setup (CR, CRLF) and other setup issues.

**Comment by rwalle61** [comment]
@GSDTN thanks for looking. I haven't had chance to do much investigation yet.

I did (hopefully) add windows support for you https://github.com/RuntimeTools/chai-openapi-response-validator/pull/32. Would you mind pulling that branch and letting me know on that PR if it works?

**Comment by GSDTN** [comment]
Thanks @rwalle61 – I was able to install the support-windows branch and run setup.test.js, definedServerPath.test.js and differentRequetModules.test.js
However, I could not execute satisfyApiSpec.test.js – not sure why, could be my IDE setup

Changing to directory test/unit/assertions and running > mocha satisfyApiSpec.test.js caused some errors because it could not resolve the 2 openapi files.

Gary

**Comment by rwalle61** [comment]
okay, I have migrated discussion about windows support to https://github.com/RuntimeTools/chai-openapi-response-validator/pull/32

**Comment by rwalle61** [comment]
@GSDTN can I see an example of the `servers` object in your OpenAPI spec? I will make sure we support your use case.

**Comment by rwalle61** [comment]
I have the code ready for various extra use cases. However, we need to decide what we should do in the 4 following cases, when the `servers` object is:

1:
```
servers: []
```

2:
```
servers:
  - url: http://api.example.com    (no base path)
```

3:
```
servers:
  - url: http://api.example.com/basePath
```

or 4:
```
servers:
  - url: /basePath
```

**In these 4 cases, should a `res.req.path` of `/basePath` match any servers?**

For case 1, [yes a `res.req.path` of `/basePath` does match a server - the default server '/'](https://swagger.io/docs/specification/api-host-and-base-path/):
>If the servers array is not provided or is empty, the server URL defaults to /:

    servers:
      - url: /


**However I cannot find instruction on what to do in cases 2-4**. What do you think/need?

**Comment by rwalle61** [comment]
I think if you don't specify a basepath (cases 1 and 2), we should assume you want to match `/`. If you do specify a basepath that does not include `/` (cases 3 and 4), we should assume you do not want to match `/`.

**Comment by GSDTN** [comment]
I suggest for matching purposes that the server portion (protocol://host.domain:port) be stripped out and only the base path portion be retained for matching.
Assume a paths:
paths:
    /myresource
    /anotherresource

Assume a request of /myresource
http://api.example.com – becomes empty space that gets prepended to the paths array values for matching – resulting in array of /myresource and /anotherresource

Assume a request of /basePath/myresource
http://api.example.com/basePath - becomes / resulting in array of /basePath/myresource and /basePath/anotherresource

**Comment by rwalle61** [comment]
Okay I will raise a PR today/tomorrow hopefully fulfilling that

**Comment by GSDTN** [comment]
I would like to add this:

This code here is causing some problems and related to the server list.
```
function getOAPath(actualRequest, openApiSpec) {
  const pathname = url.parse(actualRequest.path).pathname; // excludes the query (because: path = pathname + query)

  // needed if openApiSpec is OpenAPI 3
  const serverUrl = getServerUrlUsed(pathname, openApiSpec.servers);
  // if servers have been defined but none of the servers match the current path
  if (openApiSpec.servers && !serverUrl) {
    throw new Error(`No server matching '${pathname}' path defined in OpenAPI spec`);
  }

  const pathnameWithoutServerUrl = removeServerURLFromPath(pathname, serverUrl);
  const OAPath = findOAPathMatchingPathname(pathnameWithoutServerUrl, openApiSpec);
  return OAPath || pathnameWithoutServerUrl;
}
```

When I debug, the   `const pathname = url.parse(actualRequest.path).pathname` provides only the path portion - the server is already removed, so there is no need for the line to removeServerURLFromPath
`const pathnameWithoutServerUrl = removeServerURLFromPath(pathname, serverUrl);`

This line also suffers from the fact that in order to handle the bypass the failure I was getting with the getServerUrlUsed method, I was adding url: / to my list of servers.

This than goes though and strips off the / from the path, so /weather/hourly-observations, becomes weather/hourly-observations.
When this than tries to match to one of the path: objects which is /weather/hourly-observations, it fails.
When you make the changes to strip off the server info, that should only be to try and get the 'basePath' portion and that portion should be prefixed onto pathname.

Something similar to this:
```
function getOAPath(actualRequest, openApiSpec) {
  const pathname = url.parse(actualRequest.path).pathname; // excludes the query (because: path = pathname + query)

  // needed if openApiSpec is OpenAPI 3
  const basePath = getBasePathFromServerUsed(pathname, openApiSpec.servers);  //Returns empty string if server found but no base path.
  // if servers have been defined but none of the servers match the current path
  if (openApiSpec.servers && !serverUrl) {
    throw new Error(`No server matching '${pathname}' path defined in OpenAPI spec`);
  }

  const pathnameWithBasePath = basePath+pathname;
  const OAPath = findOAPathMatchingPathname(pathnameWithBasePath, openApiSpec);
  return OAPath || pathnameWithBasePath;
}
```

**Comment by rwalle61** [comment]
Thanks @GSDTN, I added a commit to https://github.com/RuntimeTools/chai-openapi-response-validator/pull/33 that hopefully resolves this issue. 

The code in your above comment seems to be from the published version (0.4.0), which has since been refactored in master. Nevertheless, your point stands and hopefully https://github.com/RuntimeTools/chai-openapi-response-validator/pull/33/commits/07cf5fc6507e3f795445a5f78def7f9717464ff8#diff-893001fb0428346173ba1ec24867bad6R56 achieves what you want. 

Re.
>When I debug, the const pathname = url.parse(actualRequest.path).pathname provides only the path portion - the server is already removed, so there is no need for the line to removeServerURLFromPath

You are correct that *absolute* servers are already removed, however, I think `pathname` will still contain *relative* server base paths (e.g. `/`), so there was a need to removeServerURLFromPath.

**Comment by GSDTN** [comment]
The removal of the / is currently breaking my tests and I don’t have a good way around that.
I need to add the / to the server list to get past the server check, but than the code below that removes the slash tries to compare my spec path of /weather/hourly-observation to weather/hourly-observation -which does not work.

I guest I have to wait until you have the fix for handling the full protocol://host.domain:port fix before my tests will work, since the / workaround no longer works because of the below removeServerURLFromPath

**Comment by rwalle61** [comment]
I'm confused, `removeServerURLFromPath` is old code. PR #33 should resolve the problem you describe. To test it, do you need me to release, or can you pull the PR branch?

Also, if the PR does not resolve your problem, could you point out how the OpenAPI specs in that PR differ from yours please?

**Comment by GSDTN** [comment]
Right now, I have a set of tests that is pulling the release master branch with my package.json have devDependencies including "chai-openapi-response-validator": "^0.4.0",
How would I update that to pull PR #33?

**Comment by GSDTN** [comment]
I have tried
npm install chai-openapi-response-validator#pull/33/head
npm install RuntimeTools/chai-openapi-response-validator#pull/33/head
npm install rwalle61/chai-openapi-response-validator#pull/33/head

But they all fail.

**Comment by GSDTN** [comment]
Nevermind – I figured out how to pull the pull request locally and than install the local.  That will work temporarily while I validate that it works.

**Comment by GSDTN** [comment]
So using PR #33 is working without any of my work arounds.

Thanks

**Comment by rwalle61** [comment]
Awesome, I have merged #33 and released this under [0.5.0](https://github.com/RuntimeTools/chai-openapi-response-validator/releases/tag/0.5.0).

Thank you for your comments; your feedback helped improve this package for everybody :)

Please reopen if the released version does not work as intended
